File: src/main/javacode/com/huangning/java/spark/sql/dataframe/CreateDataFrameFromRDDWithSchema.java
package com.huangning.java.spark.sql.dataframe;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

import java.util.Arrays;
import java.util.List;

public class CreateDataFrameFromRDDWithSchema {
    public static void main(String[] args) {
        /**
         * 1.自定义的类级别必须是public
         * 2.RDD转换成Daraframe会把自定义类中字段的名称按照Assci码排序
         * 3.自定义类中要实现序列化接口, 并且一定要重写get方法！！！
         */
        SparkConf conf = new SparkConf();
        conf.setAppName("RDDWithReflection").setMaster("local");
        JavaSparkContext sc = new JavaSparkContext(conf);
        SQLContext sqlContext = new SQLContext(sc);
        JavaRDD<String> lineRDD = sc.textFile("./data/person.txt");

        /**
         * 转成Row类型的RDD
         */
        final JavaRDD<Row> rowRDD = lineRDD.map(new Function<String, Row>() {
            private static final long serialVersionUID = 1;

            @Override
            public Row call(String s) throws Exception {
                return RowFactory.create(
                        s.split(",")[0],
                        s.split(",")[1],
                        Integer.valueOf(s.split(",")[2])
                );
            }
        });
        /**
         * 动态构建DataFrame中的元数据，字段可以来自字符串，也可以来自外部数据库
         */
        final List<StructField> asList = Arrays.asList(
                DataTypes.createStructField("id", DataTypes.StringType, true),
                DataTypes.createStructField("name", DataTypes.StringType, true),
                DataTypes.createStructField("age", DataTypes.IntegerType, true)
        );

        StructType schema = DataTypes.createStructType(asList);

        Dataset<Row> dataFrame = sqlContext.createDataFrame(rowRDD, schema);
        dataFrame.show();
    }
}

File: src/main/scala/com/huangning/spark/sql/dataframe/CreateDataFrameFromRDDWithSchema.scala
package com.huangning.spark.sql.dataframe

import org.apache.spark.rdd.RDD
import org.apache.spark.sql.types._
import org.apache.spark.sql.{DataFrame, Row, RowFactory, SparkSession}

/**
 * Author: 黄宁
 * Date:  2020/06/15
 * Target: 通过动态创建Schema的方式加载DataFrame
 * Note：动态创建的ROW中数据的顺序要与Schema中的数据保持一致
 */
object CreateDataFrameFromRDDWithSchema {
  def main(args: Array[String]): Unit = {
    val spark: SparkSession = SparkSession.builder().master("local").appName("createdataframefromrddwithschema").getOrCreate()
    val peopleRDD: RDD[String] = spark.sparkContext.textFile("./data/people.txt")

    // 将peopleRDD转换成RDD[Row]
    val rowRDD: RDD[Row] = peopleRDD.map(line => {
      val arr: Array[String] = line.split(",")
      Row(arr(2).toInt, arr(1), arr(0).toInt, arr(3).toLong)
    })

    // 可以动态地里面塞入列
    val structType: StructType = StructType(List[StructField](
      StructField("id", IntegerType, nullable = true),
      StructField("name", StringType, nullable = true),
      StructField("age", IntegerType, nullable = true),
      StructField("score", LongType, nullable = true)
    ))

    // createDataFrame将二者自动映射
    val frame: DataFrame = spark.createDataFrame(rowRDD, structType)
    frame.show()
    frame.printSchema()
    frame.createOrReplaceTempView("people")

    //    val schemaString = "id name age score"
    //    /**
    //      * 动态创建Schema方式
    //      */
    //    val fields: ArrayNote[StructField] = schemaString.split(" ").map(fieldName => StructField(fieldName, StringType, nullable = true))
    //    val schema: StructType = StructType(fields)
    //
    //    val rowRDD: RDD[Row] = peopleRDD
    //      .map(_.split(","))
    //      .map(attributes => Row(attributes(0).trim, attributes(1).trim, attributes(2).trim, attributes(3).trim))

    import spark.implicits._
    val results: DataFrame = spark.sql("SELECT name FROM people")
    results.map(teenager => "Name: " + teenager(0)).as("myCol").show()

  }
}
